src/Adapter/CachingAdapter.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Laminas\Paginator\Adapter;
+
+use DateInterval;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Cache\InvalidArgumentException;
+
+use function sprintf;
+
+/**
+ * @template TKey of array-key
+ * @template TValue
+ * @implements AdapterInterface<TKey, TValue>
+ */
+final readonly class CachingAdapter implements AdapterInterface
+{
+    /**
+     * @param AdapterInterface<TKey, TValue> $adapter
+     * @param non-empty-string $cacheKeyPrefix
+     */
+    public function __construct(
+        private AdapterInterface $adapter,
+        private string $cacheKeyPrefix,
```

Comment:
I'm unsure about the need for this: would leave this completely to the `CacheItemPoolInterface` to prefix on its own, if necessary

Reply:
The reason it's here is because it's likely that users would use perhaps 1 specific pool for paginators, _(that they would surely namespace)_, but in an application context might use paginator for many different result sets. Without some kind of unique cache-key element, pagination results would come from the cache anywhere the page size and offset were the same. We can't do something like `md5(serialize($theResults))` to generate a uique key for several reasons, and the only other information the adapter has must be provided to the constructor.